### ISSUE TITLE
[BUG] new server: fail2ban and opendkim errors: cpu hogged

### ISSUE THREAD

**Issue by hanscees** [docker-mailserver/docker-mailserver#3085]
### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

fail2ban server fails 

### What happened and when does this occur?

mailserver  | 2023-02-11 23:18:48,020 DEBG 'fail2ban' stderr output:
mailserver  | 2023-02-11 23:18:48,018 fail2ban                [710]: ERROR   Could not start server. Maybe an old socket file is still present. Try to remove /var/run/fail2ban/fail2ban.sock. If you used fail2ban-client to start the server, adding the -x option will do it
mailserver  | 
mailserver  | 2023-02-11 23:18:53,037 DEBG 'fail2ban' stderr output:
mailserver  | /usr/local/bin/fail2ban-wrapper.sh: line 27: /var/run/fail2ban/fail2ban.pid: No such file or directory
mailserver  | 
mailserver  | 2023-02-11 23:18:53,039 DEBG fd 51 closed, stopped monitoring <POutputDispatcher at 139749306469920 for <Subprocess at 139749306469872 with name fail2ban in state RUNNING> (stdout)>
mailserver  | 2023-02-11 23:18:53,039 DEBG 'fail2ban' stderr output:
mailserver  | /usr/local/bin/fail2ban-wrapper.sh: line 27: kill: `': not a pid or valid job spec
mailserver  | 

### What did you expect to happen?

Hi, I am setting up a new server from scratch

A few problems:

1. In the environment I disabled dkim, but it still starts and crashes because a file does not exist
2. Worse, fail2ban server does not start properly. After a while there are 15 server processes eating all cpu

### How do we replicate the issue?

1. start a new server in docker
2. use my config I think , I disabled opendkim and clamav and spamassasin and enabled rspamd
3. see fail2ban fail? 
...

Also I noticed  although I have disabled opendkim it still starts it up anyway

### DMS version

v11.3.0

### What operating system is DMS running on?

Linux

### Which operating system version?

vmware photon 4

### What instruction set architecture is DMS running on?

AMD64 / x86_64

### What container orchestration tool are you using?

Docker

### docker-compose.yml

```yml
services:
  mailserver:
    image: docker.io/mailserver/docker-mailserver:latest
    container_name: mailserver
    # If the FQDN for your mail-server is only two labels (eg: example.com),
    # you can assign this entirely to `hostname` and remove `domainname`.
    hostname: mail
    domainname: fanscees.net
    env_file: mailserver.env
    # More information about the mail-server ports:
    # https://docker-mailserver.github.io/docker-mailserver/edge/config/security/understanding-the-ports/
    # To avoid conflicts with yaml base-60 float, DO NOT remove the quotation marks.
    ports:
      - "25:25"    # SMTP  (explicit TLS => STARTTLS)
      - "143:143"  # IMAP4 (explicit TLS => STARTTLS)
      - "465:465"  # ESMTP (implicit TLS)
      - "587:587"  # ESMTP (explicit TLS => STARTTLS)
      - "993:993"  # IMAP4 (implicit TLS)
      - "995:995"  # pop3 (implicit TLS)
      - "110:110"  # pop3 (implicit TLS)
    volumes:
      - etc:/etc
      - maildata:/var/mail/
      - mailstate:/var/mail-state/
      - log:/var/log/
      - Tconfig:/tmp/docker-mailserver/
      - /etc/localtime:/etc/localtime:ro
      - ssl:/tmp/ssl
    restart: always
    stop_grace_period: 1m
    environment:
     # If you need SSL connection, you can provide your own certificates
     - SSL_TYPE=manual
     - SSL_CERT_PATH=/tmp/ssl/fullchain.pem
     - SSL_KEY_PATH=/tmp/ssl/privkey.pem
    cap_add:
      - NET_ADMIN
     healthcheck:
      test: "ss --listening --tcp | grep -P 'LISTEN.+:smtp' || exit 1"
      timeout: 3s
      retries: 0

volumes:
  maildata:
  mailstate:
  Tconfig:
  log:
  ssl:
  etc:
```

### Relevant log output

```Text
mailserver  | ACCOUNT_PROVISIONER='FILE'
mailserver  | AMAVIS_LOGLEVEL='2'
mailserver  | CLAMAV_MESSAGE_SIZE_LIMIT='25M'
mailserver  | DEFAULT_RELAY_HOST=''
mailserver  | DOVECOT_INET_PROTOCOLS='ipv4'
mailserver  | DOVECOT_MAILBOX_FORMAT='maildir'
mailserver  | DOVECOT_TLS='no'
mailserver  | ENABLE_AMAVIS='0'
mailserver  | ENABLE_CLAMAV='0'
mailserver  | ENABLE_DNSBL='1'
mailserver  | ENABLE_FAIL2BAN='1'
mailserver  | ENABLE_FETCHMAIL='0'
mailserver  | ENABLE_MANAGESIEVE='0'
mailserver  | ENABLE_POP3='1'
mailserver  | ENABLE_POSTGREY='1'
mailserver  | ENABLE_QUOTAS='0'
mailserver  | ENABLE_SASLAUTHD='0'
mailserver  | ENABLE_SPAMASSASSIN='0'
mailserver  | ENABLE_SPAMASSASSIN_KAM='1'
mailserver  | ENABLE_SRS='0'
mailserver  | ENABLE_UPDATE_CHECK='1'
mailserver  | FAIL2BAN_BLOCKTYPE='drop'
mailserver  | FETCHMAIL_PARALLEL='0'
mailserver  | FETCHMAIL_POLL='300'
mailserver  | LOGROTATE_INTERVAL='weekly'
mailserver  | LOGWATCH_INTERVAL='none'
mailserver  | LOGWATCH_RECIPIENT='hanscees@hanscees.com'
mailserver  | LOGWATCH_SENDER='mailserver-report@mail.hanscees.net'
mailserver  | LOG_LEVEL='debug'
mailserver  | MOVE_SPAM_TO_JUNK='0'
mailserver  | NETWORK_INTERFACE='eth0'
mailserver  | ONE_DIR='1'
mailserver  | OVERRIDE_HOSTNAME='mail.hanscees.net'
mailserver  | PERMIT_DOCKER='none'
mailserver  | PFLOGSUMM_RECIPIENT='hanscees@hanscees.com'
mailserver  | PFLOGSUMM_SENDER='mailreport@hanscees.com'
mailserver  | PFLOGSUMM_TRIGGER='logrotate'
mailserver  | POSTFIX_INET_PROTOCOLS='ipv4'
mailserver  | POSTFIX_MAILBOX_SIZE_LIMIT='0'
mailserver  | POSTFIX_MESSAGE_SIZE_LIMIT='80240000'
mailserver  | POSTGREY_AUTO_WHITELIST_CLIENTS='3'
mailserver  | POSTGREY_DELAY='15'
mailserver  | POSTGREY_MAX_AGE='35'
mailserver  | POSTGREY_TEXT='Delayed by Postgrey'
mailserver  | POSTMASTER_ADDRESS='hanscees@hanscees.com'
mailserver  | POSTSCREEN_ACTION='enforce'
mailserver  | RELAY_HOST=''
mailserver  | REPORT_RECIPIENT='hanscees@hanscees.com'
mailserver  | REPORT_SENDER='mailserver-report@mail.hanscees.net'
mailserver  | SA_KILL='10.0'
mailserver  | SA_SPAM_SUBJECT='***SPAM*****'
mailserver  | SA_TAG2='6.31'
mailserver  | SA_TAG='2.0'
mailserver  | SMTP_ONLY='0'
mailserver  | SPAMASSASSIN_SPAM_TO_INBOX='1'
mailserver  | SPOOF_PROTECTION='1'
mailserver  | SRS_SENDER_CLASSES='envelope_sender'
mailserver  | SSL_TYPE='manual'
mailserver  | SUPERVISOR_LOGLEVEL='debug'
mailserver  | TLS_LEVEL='modern'
mailserver  | TZ='Europe/Amsterdam'
mailserver  | UPDATE_CHECK_INTERVAL='1d'
mailserver  | VIRUSMAILS_DELETE_DELAY='7'
mailserver  | 2023-02-11 22:30:14,701 DEBG 'mailserver' stdout output:
mailserver  | ACCOUNT_PROVISIONER='FILE'
mailserver  | AMAVIS_LOGLEVEL='2'
mailserver  | CLAMAV_MESSAGE_SIZE_LIMIT='25M'
mailserver  | DEFAULT_RELAY_HOST=''
mailserver  | DOVECOT_INET_PROTOCOLS='ipv4'
mailserver  | DOVECOT_MAILBOX_FORMAT='maildir'
mailserver  | DOVECOT_TLS='no'
mailserver  | ENABLE_AMAVIS='0'
mailserver  | ENABLE_CLAMAV='0'
mailserver  | ENABLE_DNSBL='1'
mailserver  | ENABLE_FAIL2BAN='1'
mailserver  | ENABLE_FETCHMAIL='0'
mailserver  | ENABLE_MANAGESIEVE='0'
mailserver  | ENABLE_POP3='1'
mailserver  | ENABLE_POSTGREY='1'
mailserver  | ENABLE_QUOTAS='0'
mailserver  | ENABLE_SASLAUTHD='0'
mailserver  | ENABLE_SPAMASSASSIN='0'
mailserver  | ENABLE_SPAMASSASSIN_KAM='1'
mailserver  | ENABLE_SRS='0'
mailserver  | ENABLE_UPDATE_CHECK='1'
mailserver  | FAIL2BAN_BLOCKTYPE='drop'
mailserver  | FETCHMAIL_PARALLEL='0'
mailserver  | FETCHMAIL_POLL='300'
mailserver  | LOGROTATE_INTERVAL='weekly'
mailserver  | LOGWATCH_INTERVAL='none'
mailserver  | LOGWATCH_RECIPIENT='hanscees@fanscees.com'
mailserver  | LOGWATCH_SENDER='mailserver-report@mail.fanscees.net'
mailserver  | LOG_LEVEL='debug'
mailserver  | MOVE_SPAM_TO_JUNK='0'
mailserver  | NETWORK_INTERFACE='eth0'
mailserver  | ONE_DIR='1'
mailserver  | OVERRIDE_HOSTNAME='mail.fanscees.net'
mailserver  | PERMIT_DOCKER='none'
mailserver  | PFLOGSUMM_RECIPIENT='hanscees@fanscees.com'
mailserver  | PFLOGSUMM_SENDER='mailreport@fanscees.com'
mailserver  | PFLOGSUMM_TRIGGER='logrotate'
mailserver  | POSTFIX_INET_PROTOCOLS='ipv4'
mailserver  | POSTFIX_MAILBOX_SIZE_LIMIT='0'
mailserver  | POSTFIX_MESSAGE_SIZE_LIMIT='80240000'
mailserver  | POSTGREY_AUTO_WHITELIST_CLIENTS='3'
mailserver  | POSTGREY_DELAY='15'
mailserver  | POSTGREY_MAX_AGE='35'
mailserver  | POSTGREY_TEXT='Delayed by Postgrey'
mailserver  | POSTMASTER_ADDRESS='hanscees@fanscees.com'
mailserver  | POSTSCREEN_ACTION='enforce'
mailserver  | RELAY_HOST=''
mailserver  | REPORT_RECIPIENT='hanscees@fanscees.com'
mailserver  | REPORT_SENDER='mailserver-report@mail.fanscees.net'
mailserver  | SA_KILL='10.0'
mailserver  | SA_SPAM_SUBJECT='***SPAM*****'
mailserver  | SA_TAG2='6.31'
mailserver  | SA_TAG='2.0'
mailserver  | SMTP_ONLY='0'
mailserver  | SPAMASSASSIN_SPAM_TO_INBOX='1'
mailserver  | SPOOF_PROTECTION='1'
mailserver  | SRS_SENDER_CLASSES='envelope_sender'
mailserver  | SSL_TYPE='manual'
mailserver  | SUPERVISOR_LOGLEVEL='debug'
mailserver  | TLS_LEVEL='modern'
mailserver  | TZ='Europe/Amsterdam'
mailserver  | UPDATE_CHECK_INTERVAL='1d'
mailserver  | VIRUSMAILS_DELETE_DELAY='7'
```

### Other relevant information

I have no idea why this happens. Two different processes complain about a missing file opendkim and fail2ban.
That is odd. 

Only volume I do not throw away is my email addresses data, so when docker image is destroyed the postfix email addresses remain in the docker volume. 

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [X] I am rather experienced with docker
- [ ] I am inexperienced with mail servers
- [X] I am rather experienced with mail servers
- [ ] I am uncomfortable with the CLI
- [X] I am rather comfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

**Comment by georglauterbach** [comment]
Hey there :) I am not sure how you disabled DKIM on v11.3.0, as this is only possible on a relatively current `:edge` tag. Please provide **the whole startup log**. Please also try upgrading to the most recent `:edge`, because Fail2Ban/Postfix have seen changes to how they are started.

Note that with 11.3.0 (and v11.3.1), you're unable to disable OpenDKIM.

**Comment by hanscees** [comment]
Uhm, by following your websites guidelines. 

DMS_GITHUB_URL='https://raw.githubusercontent.com/docker-mailserver/docker-mailserver/master'
wget "${DMS_GITHUB_URL}/docker-compose.yml"
wget "${DMS_GITHUB_URL}/mailserver.env"
wget "${DMS_GITHUB_URL}/setup.sh"
chmod a+x ./setup.sh

and then in the env file 

# Enables the OpenDKIM service.
# **1** => Enabled
#   0   => Disabled
ENABLE_OPENDKIM=0

**Comment by hanscees** [comment]
from the startuplog these are interesting

```
2023-02-12 22:04:35,815 INFO spawned: 'opendkim' with pid 801
mailserver  | 2023-02-12 22:04:35,836 DEBG 'opendkim' stderr output:
mailserver  | opendkim: /etc/opendkim.conf: refile:/etc/opendkim/TrustedHosts: dkimf_db_open(): No such file or directory
mailserver  | 
mailserver  | 2023-02-12 22:04:35,838 INFO success: opendkim entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
mailserver  | 2023-02-12 22:04:35,839 DEBG fd 29 closed, stopped monitoring <POutputDispatcher at 139846891086752 for <Subprocess at 139846891864176 with name opendkim in state RUNNING> (stdout)>
mailserver  | 2023-02-12 22:04:35,839 DEBG fd 49 closed, stopped monitoring <POutputDispatcher at 139846891088576 for <Subprocess at 139846891864176 with name opendkim in state RUNNING> (stderr)>
mailserver  | 2023-02-12 22:04:35,840 INFO exited: opendkim (exit status 78; not expected)
mailserver  | 2023-02-12 22:04:35,841 DEBG received SIGCHLD indicating a child quit
mailserver  | 2023-02-12 22:04:35,846 INFO spawned: 'opendkim' with pid 811
mailserver  | 2023-02-12 22:04:35,863 DEBG 'opendkim' stderr output:
mailserver  | opendkim: /etc/opendkim.conf: refile:/etc/opendkim/TrustedHosts: dkimf_db_open(): No such file or directory
mailserver  | 
mailserver  | 2023-02-12 22:04:35,863 INFO success: opendkim entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
mailserver  | 2023-02-12 22:04:35,864 DEBG fd 29 closed, stopped monitoring <POutputDispatcher at 139846891087088 for <Subprocess at 139846891864176 with name opendkim in state RUNNING> (stdout)>
mailserver  | 2023-02-12 22:04:35,865 DEBG fd 49 closed, stopped monitoring <POutputDispatcher at 139846891087856 for <Subprocess at 139846891864176 with name opendkim in state RUNNING> (stderr)>
mailserver  | 2023-02-12 22:04:35,866 INFO exited: opendkim (exit status 78; not expected)
mailserver  | 2023-02-12 22:04:35,866 DEBG received SIGCHLD indicating a child quit
mailserver  | 2023-02-12 22:04:35,870 INFO spawned: 'opendkim' with pid 814
mailserver  | 2023-02-12 22:04:35,886 DEBG 'opendkim' stderr output:
mailserver  | opendkim: /etc/opendkim.conf: refile:/etc/opendkim/TrustedHosts: dkimf_db_open(): No such file or directory

```

and also

```
mailserver  | 2023-02-12 21:52:58,119 DEBG 'changedetector' stdout output:
mailserver  | [  DEBUG  ]  2023-02-12 21:52:58  Changedetector is ready
mailserver  | 2023-02-12 21:57:39,533 DEBG 'fail2ban' stderr output:
mailserver  | 2023-02-12 21:57:39,532 fail2ban                [709]: ERROR   Could not start server. Maybe an old socket file is still present. Try to remove /var/run/fail2ban/fail2ban.sock. If you used fail2ban-client to start the server, adding the -x option will do it
```

I tried adding the missing files, but then realized probably something else is wrong

**Comment by hanscees** [comment]
heres the log 

[startupmail.txt](https://github.com/docker-mailserver/docker-mailserver/files/10717277/startupmail.txt)

**Comment by georglauterbach** [comment]
I see where the issue is :) Please read the README more carefully; it states:

> If you have issues, read the full README and the documentation for your version (default is edge) first before opening an issue.

and our tagging convention also states that `:latest` is the latest release, not the latest container image.

---

You're running `:latest`, which is 11.3.X. You need `:edge` though when trying to disable OpenDKIM. Please try changing the image tag, then we'll see what happens.

**Comment by polarathene** [comment]
OpenDKIM I believe runs by default until the new release is made (with changes in `:edge`). If it hasn't been given all the config files, it repeatedly crashes and `supervisord` keeps restarting it. That bug has been fixed (_but also only in `:edge`_).

Fail2Ban in `:edge` has been upgraded to a newer version that won't run into the problem you experienced either. It should cleanup that socket file itself now and not end up running so many copies as the command to run it with `supervisord` has been changed / improved.

Sorry about those issues, we should hopefully have the new breaking release v12 out this month :+1: 

If you'd like to test the latest changes use the `:edge` image, that follows our master branch and will be updated each time (_where you'll have to pull the image again explicitly to update it_).

---

Also, I am fairly sure I recognize your username as someone who contributed a script to the Github Wiki before we migrated the docs back in early 2021? You're probably aware by now, but there was a [breaking change made back in `10.2` that impacted your script](https://docker-mailserver.github.io/docker-mailserver/v11.3/config/security/ssl/#importing-certificates-obtained-via-another-source) that interacted with internal-only cert files :sweat_smile: (_breaking change allowed in this sense since it was internal information, not something to be publicly relied on_)

[This is the script I remembered](https://github.com/hanscees/dockerscripts/blob/a9c74f8fbfb0b664275288a3ba9cf92a714338c1/scripts/tomav-renew-certs). If you still use that or the other ones for DMS, you may want to review them.

**Comment by georglauterbach** [comment]
Just FYI for fellow Rdpamd user: You should see significant improvements to the Rspamd setup when #3059 is merged. If you have feedback for me, just ping me or open another issue :)

**Comment by hanscees** [comment]
Apparently I missed some dropdown in your docs 😩 and if you miss that you are directed to the edge documentation. 


So to summarize if I want rspamd (I do) at this time:
1. I need to use the edge release ? It means to change the image tag from :latest to : edge in the docker-compose.yml file
2. In the documentation I had already (unknowingly) chosen edge. 



Hopefully my clumsiness didn't cost too much time. I was living on the edge but didn't know it 😁

**Comment by georglauterbach** [comment]
> Apparently I missed some dropdown in your docs 😩 and if you miss that you are directed to the edge documentation.

We do out best in the `README` :) The version selection menu is not too big though 😆

> So to summarize if I want rspamd (I do) at this time:
> 
>     1. I need to use the edge release ? It means to change the image tag from :latest to : edge in the docker-compose.yml file

Exactly! Please give me some feedback after #3059 is merged (which will require you to pull the very latest `:edge` again) on the Rspamd setup (what you think could be improved, what is already nice, etc.) :) I'd love to hear it ❤️

>     2. In the documentation I had already (unknowingly) chosen edge.

Correct as well.

> Hopefully my clumsiness didn't cost too much time. I was living on the edge but didn't know it 😁

Happens 🤣

**Comment by hanscees** [comment]
I have now red the readme page and I still don't understand howto find the right version if I want to use the stable version most users will want to use. I know it is logical to you, simnce you all live here. But my recommendation would be:

1. make the documentation on your readme page suitable for the stable version. not for the edge version
2. second best if you think thats nonsense (and most people who are in a project for a while don't see what new users see ) on your landing page
 `https://github.com/docker-mailserver/docker-mailserver`
Below the heading " get the tools"  state 
`The commands below will download the edge branch, if you want the stable branch, see the documentation.`

**Comment by georglauterbach** [comment]
Probably a good idea to add a note or two. I'd go for something similar to the second proposal. I will provide an update after my current PRs are merged.

**Comment by hanscees** [comment]
I  am running the :edge thing now and it does not hog the cpu so it looks good :)

Still wrestling a bit with the iptables and nftables rules on the vmware host the docker is running on here. Got that running, but later on its critical the fail2ban indeed bans IP's effectively. 
Looking good for now, more tomorrow.

**Comment by georglauterbach** [comment]
What's the status here?:)

**Comment by hanscees** [comment]
I cant test properly because, although your collegue closed my ticket, edge still does not copy in extentions into the config. 

```
docker exec -it mailserver sh
# 
# cd /tmp	
# ls
docker-mailserver  docker-mailserver-config-chksum  ssl
# ls do*
docker-mailserver-config-chksum

docker-mailserver:
postfix-accounts.cf  postfix-virtual.cf
# exit
root@mail [ ~/containers/mailserver ]# pwd
/root/containers/mailserver
root@mail [ ~/containers/mailserver ]# ls
docker-compose.yml  docker-data  mailserver.env  setup.sh  surbl-whitelist.inc.zst
root@mail [ ~/containers/mailserver ]# ls docker-data/dms/
config
root@mail [ ~/containers/mailserver ]# ls docker-data/dms/config/
postfix-accounts.cf  postfix-main.cf
```

and because you also made it impossible to go into the contained and fix things and just restart postfix I am stuck as a brick

**Comment by georglauterbach** [comment]
> Wait, I have found where this goes wrong, my mistake.
> 
> 
> 
> its in the docker-config file 
> 
>       - ./docker-data/dms/config/:/tmp/docker-mailserver/
> 
> 

Wrong issue? Was this supposed to go into the other issue of yours?

**Comment by hanscees** [comment]
I now have the edge mailserver in production. 
Are there specific things you want to know about it? I wasted some of your time, so you can waste some of mine too :)

**Comment by georglauterbach** [comment]
> I now have the edge mailserver in production. 

👍🏼 You will probably receive an update notification when v12.0.0 is released; for production, we usually recommend going with a pinned version, if possible.

> Are there specific things you want to know about it? I wasted some of your time, so you can waste some of mine too :) 

Not necessary, but thanks for the offer :) If there is something, we might come back to it, but for now, all is well 🙂

I will close this as completed.
